python_modules/dagster/dagster/_core/definitions/result.py

```diff
@@ -31,11 +32,15 @@ def __new__(
         check_results: Optional[Sequence[AssetCheckResult]] = None,
         data_version: Optional[DataVersion] = None,
         tags: Optional[Mapping[str, str]] = None,
+        value: Optional[Any] = None,
     ):
         from dagster._core.definitions.events import validate_asset_event_tags
 
         asset_key = AssetKey.from_coercible(asset_key) if asset_key else None
 
+        if cls is ObserveResult and value is not None:
```

Comment:
this bit is a little weird / footgunny to me -- if these classes were `@records` then we could pretty easily make it so that the base `AssetResult` class didn't even have a value property (and only `MaterializeResult` would), which might make things a bit more intuitive.